data/blog/tailscale-sucks.mdx

```diff
@@ -0,0 +1,456 @@
+  />
+  <figcaption>
+    The lidar and front sensor cluster on the Z10 Pro. The robot shoots lasers out of the angled
+    apertures, and an IR camera cluster in the center uses distortion in the beam to sense
+    obstacles.
+  </figcaption>
+</figure>
+
+However, the modern crop of robot vacuums are also firmly IoT cloud devices.
+They connect to a faraway control plane, you get a mobile app (with a ton of
+lag, since every button press is bouncing through the cloud and back to the
+bot), and the machine is a completely opaque appliance. You bought it, but as
+far as the robot is concerned, you're just a guest in its brain, invited there
+by its makers.
+
+The Z10 Pro is part of the [Xiaomi Smart Home][mi-home] ecosystem, one of the
+world's largest IoT clouds. Unfortunately, this ecosystem's empirical security
+and privacy posture is not something I'm comfortable having in my home: as
+illustrated in a [DEFCON talk][giese-talk] a few years ago, the manufacturers
+have effectively unlimited control over their robots in the field, up to and
+including spawning a remote shell and uploading raw sensor imagery. Clouds don't
+_have_ to be this invasive, but that's the reality for this robot.
+
+And really, for this application, the cloud is just getting in the way. I wish I
+could just control the robot directly somehow. And guess what, there is a way!
+Let's jailbreak my vacuum cleaner.
+
+<figure className="float-none text-center sm:pl-4">
+  <Image
+    src="/images/tailscale-sucks/z10-pre.jpg"
+    width="636"
+    height="465"
+    alt="Photo of the Z10 Pro vacuum robot sitting on a desk. a red circuit board with wires sticking out of it is partially visible on the desk to the left of the robot."
+  />
+  <figcaption>The patient, unaware of what's coming.</figcaption>
+</figure>
+
+Before we continue, I want to give credit and praise to [Dennis
+Giese][giese-site] and [Sören Beye][beye-site] for making all this possible.
+Dennis has been researching the security of IoT systems for years, and has
+compiled an encyclopedia of methods and expertise on gaining access to these
+appliances. Sören built on top of this work and created [Valetudo][valetudo].
+Valetudo is an open-source control plane for vacuum robots that runs on the
+robot itself and enables cloud-free operation, as well as integration into
+open-source automation ecosystems like [Home Assistant][hass].
+
+## Jailbreaking a vacuum cleaner
+
+I'm going to repeat the words of warning on Valetudo's website: jailbreaking
+your vacuum can be fun and isn't _too_ difficult, but you need to take your
+time, read and follow the instructions carefully, and accept the fact that this
+does carry some risk of bricking your device. This is a necessarily somewhat
+messy process, so please be aware of the risks, take your time, read the
```

Comment:
"a necessarily" reads a bit awkwardly to me.

Maybe "This is necessarily somewhat of a messy process" ?

Reply:
Done.